internal/spoke_k8s_client/factory.go

```diff
@@ -1,90 +1,255 @@
+		csrClient:   clientSet.CertificatesV1().CertificateSigningRequests(),
+		sarClient:   clientSet.AuthorizationV1().SelfSubjectAccessReviews(),
+		nodesClient: clientSet.CoreV1().Nodes(),
 	}
-	return kubeconfigData, nil
+	return
 }
 
-func (cf *spokeK8sClientFactory) clientAndSetForKubeconfig(kubeconfig []byte) (SpokeK8sClient, *kubernetes.Clientset, error) {
-	clientConfig, err := clientcmd.NewClientConfigFromBytes(kubeconfig)
+func (f *spokeK8sClientFactory) restConfigFromSecret(secret *corev1.Secret) (result *rest.Config, err error) {
+	kubeConfig, err := f.kubeConfigFromSecret(secret)
 	if err != nil {
-		return nil, nil, errors.Wrapf(err, "failed to get clientconfig from kubeconfig data")
+		return
 	}
-
-	restConfig, err := clientConfig.ClientConfig()
+	clientConfig, err := clientcmd.NewClientConfigFromBytes(kubeConfig)
 	if err != nil {
-		return nil, nil, errors.Wrapf(err, "failed to get restconfig for kube client")
+		return
 	}
+	result, err = clientConfig.ClientConfig()
+	return
+}
 
-	clientset, err := kubernetes.NewForConfig(restConfig)
-	if err != nil {
-		cf.log.WithError(err).Warnf("Getting kuberenetes config for cluster")
-		return nil, nil, err
+func (f *spokeK8sClientFactory) kubeConfigFromSecret(secret *corev1.Secret) (result []byte, err error) {
+	if len(secret.Data) == 0 {
+		err = errors.Errorf(
+			"secret '%s/%s' is empty",
+			secret.Namespace, secret.Name,
+		)
+		return
+	}
+	result, ok := secret.Data["kubeconfig"]
+	if !ok || len(result) == 0 {
+		err = errors.Errorf(
+			"secret '%s/%s' doesn't contain the 'kubeconfig' key",
+			secret.Namespace, secret.Name,
+		)
 	}
+	return
+}
 
+func (f *spokeK8sClientFactory) clientFromRestConfig(restConfig *rest.Config) (result ctrlclient.Client, err error) {
 	schemes := GetKubeClientSchemes()
-	targetClient, err := client.New(restConfig, client.Options{Scheme: schemes})
-	if err != nil {
-		cf.log.WithError(err).Warnf("failed to get spoke kube client")
-		return nil, nil, err
+	result, err = ctrlclient.New(restConfig, ctrlclient.Options{Scheme: schemes})
+	return
+}
+
+func (f *spokeK8sClientFactory) clientSetFromRestConfig(restConfig *rest.Config) (result *kubernetes.Clientset,
+	err error) {
+	result, err = kubernetes.NewForConfig(restConfig)
+	return
+}
+
+// isFromHostedCluster determines if this cluster corresponding to the given kubeconfig secret is a hosted cluster.
+// Returns a boolean flag and, when that flag is true, the namespace and name of the cluster deployment of the hosted
+// cluster.
+func (f *spokeK8sClientFactory) isFromHostedCluster(ctx context.Context, kubeconfigSecret *corev1.Secret) (result bool,
+	clusterDeploymentKey types.NamespacedName, err error) {
+	// Try to find the cluster deployment. If we can't, for whatever the reason, explain it in the log and assume
+	// it isn't a hosted cluster.
+	clusterDeployment, err := f.findClusterDeploymentForKubeconfigSecret(ctx, kubeconfigSecret)
+	if err != nil || clusterDeployment == nil {
```

Comment:
Error and not having clusterDeployment seems to be different issues, maybe we should split the logging at least?

Reply:
OK, will do.